scripts/base/utils/addrs.zeek

```diff
@@ -83,16 +83,29 @@ function has_valid_octets(octets: string_vec): bool
 ##
 ## input: a string that may contain an IP address anywhere within it.
 ##
+## check_wrapping: if true, will only return IP addresses that are wrapped in matching
+## pairs of spaces, square brackets, curly braces, or parens. This can be used to avoid
+## extracting strings that look like IPs from innocuous strings, such as SMTP headers.
+##
 ## Returns: an array containing all valid IP address strings found in *input*.
-function extract_ip_addresses(input: string): string_vec
+function extract_ip_addresses(input: string, check_wrapping: bool &default=F): string_vec
 	{
 	local parts = split_string_all(input, ip_addr_regex);
 	local output: string_vec;
 
 	for ( i in parts )
 		{
 		if ( i % 2 == 1 && is_valid_ip(parts[i]) )
-			output += parts[i];
+			{
+			local p1 = parts[i-1];
+			local p3 = parts[i+1];
```

Comment:
what happens if there is no parts[i+2]?

Reply:
`i+2`, or `i+1`?

Reply:
i+1. I was really just curious if it is possible that parts[i+1] is not defined.

Reply:
Oh - yes, I see - i messed up my comment :). i+1, sorry.

Reply:
I'm not sure how to make it actually give me a result with an even number of parts, based on the call to `split_string_all` and the specific pattern being passed to it. I tried this set of strings:

```
local a = "[1.1.1.1]";
local b = "1.1.1.1]";
local c = "[1.1.1.1";
local d = "[1.1.1.1] [1.1.1.1";
local e = "[1.1.1.1] [2.2.2.2]";
```

The first three all give me 3 parts. The 4th gives me 5. Calling `print extract_ip_addresses(d, T); print extract_ip_addresses(e, T);`, the middle `] [` is considered a single element, so the IPs are elements 1 and 3 in the parts array, which matches the check in the if statement before that block.

Reply:
I do wonder if the last condition should `( p1[-1] == " " && ( p3[0] == " " || |p3| == 0 ) ) )` though, so that a string can end with an IP address separated from the rest of the text by a space.

Reply:
I would suggest just adding general checks to ensure `i` is in the expect range before doing any of the tests (e.g., `i > 0 && i < |parts| - 1`, etc.). Even if it seemingly can't happen, it would make the code more readable because one doesn't need to wonder.